#### 24. LLM Tests Mistralv3

# Import

In [1]:
import pandas as pd

# PICKLE and Load Variables

In [2]:
import pickle
def pkl_vars(varname, filename):
    with open(filename, 'wb') as file:
        pickle.dump(varname, file)
        
def reload_vars(filename):
    this_var = None
    with open(filename, 'rb') as file:
        this_var = pickle.load(file)      
    return this_var

In [3]:
df_test_term = reload_vars('/home/ann/fiz-ddb/notebook/data/vars/flattened_test_new_term_disam_list.pkl')
df_test_term[:10]

[{'title': 'Definition:Baire Space (Topology)',
  'term': 'Baire Space',
  'definition': 'Let T = ( S, τ) be a topological space.\n\n\n=== Definition 1 ===\nLet T = ( S, τ) be a topological space.\n\n\nT is a Baire space  if and only if  the union of any countable set of closed sets of T whose interiors are empty also has an empty interior.\n\n=== Definition 2 ===\nLet T = ( S, τ) be a topological space.\n\n\nT is a Baire space  if and only if  the intersection of any countable set of open sets of T which are everywhere dense is everywhere dense.\n\n=== Definition 3 ===\nLet T = ( S, τ) be a topological space.\n\n\nT is a Baire space  if and only if  the interior of the union of any countable set of closed sets of T which are nowhere dense is empty.\n\n=== Definition 4 ===\nLet T = ( S, τ) be a topological space.\n\n\nT is a Baire space  if and only if , whenever the union of any countable set of closed sets of T has an interior point, then one of those closed sets must have an interio

In [4]:
df_test_title = reload_vars("/home/ann/fiz-ddb/notebook/data/vars/flattened_test_new_candi_disam_list.pkl")
df_test_title[:10]

[{'title': 'Definition:Quintal System',
  'term': 'System',
  'definition': 'A quintal system  is a number system in which the base is 5.\n\n\n=== Quintal Notation ===\nQuintal notation is the positional number system whose base is 5.\n\nThat is, every number x ∈ℝ is expressed in the form:\n:x = ∑_j ∈ℤ r_j 5^j\nwhere:\n:∀ j ∈ℤ: r_j ∈{ 0, 1, 2, 3, 4 }'},
 {'title': 'Definition:Decimal System',
  'term': 'System',
  'definition': 'A decimal system  is a number system in which the base is 10.\n\n\n=== Decimal Notation ===\nDecimal notation is the quotidian technique of expressing numbers in base 10.\n\nEvery number x ∈ℝ is expressed in the form:\n:x = ∑_j ∈ℤ r_j 10^j\nwhere:\n:∀ j ∈ℤ: r_j ∈{ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 }\n\n=== Decimal System of Measurement ===\nA decimal system (of measurement) is a system of measurement in which the standard multiples and fractions of the units of measurement are powers of 10.'},
 {'title': 'Definition:Vigesimal System',
  'term': 'System',
  'definiti

In [14]:
df_term_kb = reload_vars('/home/ann/fiz-ddb/notebook/data/vars/def_disamb_kb.pkl')
df_term_kb["Loop"]

['Definition:Loop (Plane Geometry)',
 'Definition:Loop (Graph Theory)',
 'Definition:Algebra Loop',
 'Definition:Relational Loop',
 'Definition:Loop (Topology)',
 'Definition:Loop (Matroid)']

# Save Variables

https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3

# Mistral v3

In [8]:
!pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu124

Looking in indexes: https://download.pytorch.org/whl/cu124
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 90.7 MB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 60.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 74.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 59.8 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 70.3 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 75.1 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 75.0 MB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 72.5 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 72.4 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 188.7/188.7 MB 94.6 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21

In [6]:
!nvidia-smi

Sat Mar  1 13:04:16 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.14              Driver Version: 550.54.14      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A100 80GB PCIe          Off |   00000000:21:00.0 Off |                    0 |
| N/A   35C    P0             45W /  300W |       3MiB /  81920MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+-----

In [9]:
!pip install mistral_inference

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.5/6.5 MB 36.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.3/43.3 MB 79.8 MB/s eta 0:00:00ta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 85.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 81.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 67.7 MB/s eta 0:00:00
doneheel for fire (setup.py) ... 
  Created wheel for fire: filename=fire-0.7.0-py3-none-any.whl size=114298 sha256=d7ac6238a92169975f4a2ec7cd6653ce7fba6924557efef6ca868eb6f6219632
  Stored in directory: /home/ann/.cache/pip/wheels/19/39/2f/2d3cadc408a8804103f1c34ddd4b9f6a93497b11fa96fe738e
Successfully built fire


In [10]:
## Setup Device

In [3]:
import torch
torch.cuda.is_available()
device = "cuda:0" if torch.cuda.is_available() else "cpu"
device

'cuda:0'

In [12]:
!pip install transformers

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.0/10.0 MB 45.0 MB/s eta 0:00:00 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 91.2 MB/s eta 0:00:00


In [14]:
!pip install huggingface_hub

In [22]:
from huggingface_hub import notebook_login
notebook_login()

## Test zero-shot

From https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3

https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3

In [4]:
from huggingface_hub import snapshot_download
from pathlib import Path

mistral_models_path = Path.home().joinpath('mistral_models', '7B-Instruct-v0.3')
mistral_models_path.mkdir(parents=True, exist_ok=True)

snapshot_download(repo_id="mistralai/Mistral-7B-Instruct-v0.3", allow_patterns=["params.json", "consolidated.safetensors", "tokenizer.model.v3"], local_dir=mistral_models_path)

Fetching 3 files:   0%|          | 0/3 [00:00<?, ?it/s]

'/home/ann/mistral_models/7B-Instruct-v0.3'

In [5]:
setting = "Your task is to find the correct article title given a mathematical term and definition as context.\n"

In [6]:
no_answer = "Reply with \"I don\'t know\" when uncertain. \n"

In [7]:
no_explanation = "Only select one answer from the provided list. \nDo not provide justifications.\n"

In [8]:
definition = "\n\nFor example, the endpoints of a line segment are its boundaries.\n\n\n=== Containment ===\nA geometric figure is said to be contained by its boundary or boundaries.\n\n "

In [9]:
term = "\"Boundary\""

In [10]:
choices = ['Boundary (Geometry)',
  'Boundary (Topology)',
  'Boundary (Graph Theory)']

In [11]:
test_inst = setting + no_answer + no_explanation + \
            "Identify the correct definition title from this list: \n" + str(choices) + \
            "\n\nGiven the term " + term + \
            " and the following mathematical definition as context: \n" + definition
print(test_inst)

Your task is to find the correct article title given a mathematical term and definition as context.
Reply with "I don't know" when uncertain. 
Only select one answer from the provided list. 
Do not provide justifications.
Identify the correct definition title from this list: 
['Boundary (Geometry)', 'Boundary (Topology)', 'Boundary (Graph Theory)']

Given the term "Boundary" and the following mathematical definition as context: 


For example, the endpoints of a line segment are its boundaries.


=== Containment ===
A geometric figure is said to be contained by its boundary or boundaries.

 


### Instruct following test

In [12]:
from mistral_inference.transformer import Transformer
from mistral_inference.generate import generate

from mistral_common.tokens.tokenizers.mistral import MistralTokenizer
from mistral_common.protocol.instruct.messages import UserMessage
from mistral_common.protocol.instruct.request import ChatCompletionRequest


tokenizer = MistralTokenizer.from_file(f"{mistral_models_path}/tokenizer.model.v3")
model = Transformer.from_folder(mistral_models_path)

completion_request = ChatCompletionRequest(messages=[UserMessage(content=test_inst)])

tokens = tokenizer.encode_chat_completion(completion_request).tokens

out_tokens, _ = generate([tokens], model, max_tokens=255, temperature=0.0, eos_id=tokenizer.instruct_tokenizer.tokenizer.eos_id)
result = tokenizer.instruct_tokenizer.tokenizer.decode(out_tokens[0])

print(result)

'Boundary (Geometry)'


In [15]:
def get_choices(term):
    global df_term_kb
    choices = list()
    for idx, choice in enumerate(df_term_kb[term]):
        #text = f'{idx+1}. {choice.split(":")[-1]}'        
        text = f'{choice.split(":")[-1]}'        
        choices.append(text)
    
    return choices    
    

get_choices("Boundary")

['Boundary (Geometry)', 'Boundary (Topology)', 'Boundary (Graph Theory)']

In [25]:
len(df_test_term)

364

In [26]:
len(df_test_title)

462

In [16]:
from rich.progress import track, Progress
from rich import print
import datetime, time

def run_test(data, logfile, offset):

    global tokenizer
    global model
    
    answers = list()
    total_time = 0
    messages = list()

    print(datetime.datetime.now())

    with Progress() as progress:

        task1 = progress.add_task("[blue] Prompt sample [%s]" % (len(data)-offset), total=len(data)-offset)

        with open(logfile, 'w') as f_logs:

            for idx, elem in enumerate(data[offset:]):
                progress.update(task1, advance=1)
    
                choices = get_choices(elem['term'])
                inst = setting + \
                    no_answer + \
                    no_explanation + \
                    "Identify the correct definition title from this list: \n" + str(choices) +\
                    "\nGiven the term " + elem['term'] + \
                    "and the following mathematical definition as context: \n" + elem['definition']
    
                #messages.append([{"role": "user", "content": inst}])
                messages = [{"role": "user", "content": inst}]
    
                #print(f'start [{idx}]: {datetime.datetime.now()}')
                start = time.time()

                completion_request = ChatCompletionRequest(messages=[UserMessage(content=inst)])

                tokens = tokenizer.encode_chat_completion(completion_request).tokens

                out_tokens, _ = generate([tokens], model, max_tokens=255, temperature=0.0, eos_id=tokenizer.instruct_tokenizer.tokenizer.eos_id)
                result = tokenizer.instruct_tokenizer.tokenizer.decode(out_tokens[0])
    
                end = time.time()
                elapsed_time = end - start
                total_time += elapsed_time
                logstr = f'[{idx + offset}],{elapsed_time}\n'
                f_logs.write(logstr)
    
                answers.append((idx, elem['definition'].split(':')[-1], result))        

        return total_time, answers
        

## test_term

In [31]:
test_term_time, test_term_answers = run_test(df_test_term, logfile='/home/ann/fiz-ddb/notebook/data/mistralv3-exec-01-test-term.csv', offset=0)

2025-03-01 21:31:51.007708

Output()

In [32]:
print(test_term_time)
print(len(test_term_answers))

206.76170134544373

364

In [33]:
test_term_answers[:5]

[(0,
  'Let T = ( S, τ) be a topological space.\n\n\n=== Definition 1 ===\nLet T = ( S, τ) be a topological space.\n\n\nT is a Baire space  if and only if  the union of any countable set of closed sets of T whose interiors are empty also has an empty interior.\n\n=== Definition 2 ===\nLet T = ( S, τ) be a topological space.\n\n\nT is a Baire space  if and only if  the intersection of any countable set of open sets of T which are everywhere dense is everywhere dense.\n\n=== Definition 3 ===\nLet T = ( S, τ) be a topological space.\n\n\nT is a Baire space  if and only if  the interior of the union of any countable set of closed sets of T which are nowhere dense is empty.\n\n=== Definition 4 ===\nLet T = ( S, τ) be a topological space.\n\n\nT is a Baire space  if and only if , whenever the union of any countable set of closed sets of T has an interior point, then one of those closed sets must have an interior point.',
  "'Baire Space (Topology)'"),
 (1,
  'The Baire space 𝐁 is defined as 

## test_title

# Test run

In [73]:
len(df_test_title)

462

In [74]:
test_title_time, test_title_answers = run_test(df_test_title, logfile='/home/ann/fiz-ddb/notebook/data/mistralv3-exec-01-test-title.csv', offset=0)

2025-03-01 21:59:43.468968

Output()

# Analysis

## test_term

In [34]:
len(test_term_answers)

364

{'Boundary': ['Definition:Boundary (Geometry)',
  'Definition:Boundary (Topology)',
  'Definition:Boundary (Graph Theory)'],
 'Right': ['Definition:Right (Direction)',
  'Definition:Right Angle',
  'Definition:Right Hand Side',
  'Definition:Right-Total Relation',
  'Definition:Right Quasi-Reflexive Relation`

In [36]:
import re

def get_results(df_test, raw_answers):
    results = list()
    correct_count = 0
    
    for idx, elem in enumerate(zip(df_test, raw_answers)):
        (gt, answer) = elem
        choices = get_choices(gt['term'])
        gt = gt['title'].split('Definition:')[-1]
        
        is_correct = False
        answer = re.sub("[\"\']", "", answer[2])
        
        #if gt.lower().strip() in answer[1].lower().strip(): #lax
        #if gt in answer or answer in gt:
        if answer.lower().strip() in gt.lower().strip():

            is_correct = True
            correct_count += 1

        results.append({
            'idx': idx,
            'is_correct': is_correct,
            'gt': gt,
            'pred': answer,
            'choices': choices
        })

    return correct_count, results

In [67]:
test_term_correct_count, test_term_results = get_results(df_test_term, test_term_answers)

### Scores

In [68]:
print(f'Accuracy: {test_term_correct_count/len(df_test_term)*100:0.2f}')

Accuracy: 67.03

In [69]:
pkl_vars(test_term_results, '/home/ann/fiz-ddb/notebook/data/vars/mistralv3-results-01-test-term.pkl')

## test_title

In [75]:
test_title_correct_count, test_title_results = get_results(df_test_title, test_title_answers)

In [76]:
print(f'Accuracy: {test_title_correct_count/len(df_test_title)*100:0.2f}')

Accuracy: 69.05

In [77]:
pkl_vars(test_title_results, '/home/ann/fiz-ddb/notebook/data/vars/mistralv3-results-01-test-title.pkl')

## Error Analysis

In [18]:
def get_errors(results):
    incorrect = list()
    
    for result in results:
        if not result['is_correct']:
            incorrect.append(result)
    return incorrect

In [71]:
test_term_incorrect = get_errors(test_term_results)

In [55]:
len(test_term_incorrect)

120

In [72]:
import json

with open('/home/ann/fiz-ddb/notebook/data/mistralv3-errors-01-test-term.txt', 'w') as f:
    f.write(json.dumps(test_term_incorrect, indent=4, sort_keys=True))

In [79]:
test_title_incorrect = get_errors(test_title_results)

with open('/home/ann/fiz-ddb/notebook/data/mistralv3-errors-01-test-title.txt', 'w') as f:
    f.write(json.dumps(test_title_incorrect, indent=4, sort_keys=True))

# Cleanup

In [447]:
import torch, gc
model = None
tokenizer = None
encodeds = None

gc.collect()

del encodeds
del tokenizer
del model

torch.cuda.empty_cache()

# 5-fold

In [19]:
test_files = ["flattened_test_new_candi_1_disam_list.pkl", \
                    "flattened_test_new_candi_2_disam_list.pkl", \
                    "flattened_test_new_candi_3_disam_list.pkl", \
                    "flattened_test_new_candi_4_disam_list.pkl", \
                    "flattened_test_new_candi_5_disam_list.pkl", \
                    "flattened_test_new_term_1_disam_list.pkl", \
                    "flattened_test_new_term_2_disam_list.pkl", \
                    "flattened_test_new_term_3_disam_list.pkl", \
                    "flattened_test_new_term_4_disam_list.pkl", \
                    "flattened_test_new_term_5_disam_list.pkl"]
print(test_files)

[
    'flattened_test_new_candi_1_disam_list.pkl',
    'flattened_test_new_candi_2_disam_list.pkl',
    'flattened_test_new_candi_3_disam_list.pkl',
    'flattened_test_new_candi_4_disam_list.pkl',
    'flattened_test_new_candi_5_disam_list.pkl',
    'flattened_test_new_term_1_disam_list.pkl',
    'flattened_test_new_term_2_disam_list.pkl',
    'flattened_test_new_term_3_disam_list.pkl',
    'flattened_test_new_term_4_disam_list.pkl',
    'flattened_test_new_term_5_disam_list.pkl'
]

In [38]:
TEST_DATA = {}

In [41]:
import copy

for filename in test_files:
    fullpath = f'/home/ann/fiz-ddb/notebook/data/vars/{filename}'
    
    df_test = reload_vars(fullpath)
    logfile = f'/home/ann/fiz-ddb/notebook/data/5fold-mistralv3/{filename}.log'

    print(fullpath, logfile)
    
    test_time, test_answers = run_test(df_test, logfile=logfile, offset=0)

    test_correct_count, test_results = get_results(df_test, test_answers)

    TEST_DATA[filename] = {
        'compute_time': test_time,
        'logfile': logfile,
        'raw_answers': copy.deepcopy(test_answers),
        'correct_count': test_correct_count,
        'test_results': copy.deepcopy(test_results)
    }

    print(f'[{test_correct_count/len(df_test)*100:0.2f}] ({filename})')
    pkl_vars(TEST_DATA, '/home/ann/fiz-ddb/notebook/data/vars/mathd2_5fold_mistralv3_20250310.pkl')

/home/ann/fiz-ddb/notebook/data/vars/flattened_test_new_candi_1_disam_list.pkl 
/home/ann/fiz-ddb/notebook/data/5fold-mistralv3/flattened_test_new_candi_1_disam_list.pkl.log

2025-03-10 14:35:07.466313

Output()

[50.60] (flattened_test_new_candi_1_disam_list.pkl)

/home/ann/fiz-ddb/notebook/data/vars/flattened_test_new_candi_2_disam_list.pkl 
/home/ann/fiz-ddb/notebook/data/5fold-mistralv3/flattened_test_new_candi_2_disam_list.pkl.log

2025-03-10 14:39:20.831251

Output()

[52.15] (flattened_test_new_candi_2_disam_list.pkl)

/home/ann/fiz-ddb/notebook/data/vars/flattened_test_new_candi_3_disam_list.pkl 
/home/ann/fiz-ddb/notebook/data/5fold-mistralv3/flattened_test_new_candi_3_disam_list.pkl.log

2025-03-10 14:43:44.845940

Output()

[50.54] (flattened_test_new_candi_3_disam_list.pkl)

/home/ann/fiz-ddb/notebook/data/vars/flattened_test_new_candi_4_disam_list.pkl 
/home/ann/fiz-ddb/notebook/data/5fold-mistralv3/flattened_test_new_candi_4_disam_list.pkl.log

2025-03-10 14:48:23.233414

Output()

[54.88] (flattened_test_new_candi_4_disam_list.pkl)

/home/ann/fiz-ddb/notebook/data/vars/flattened_test_new_candi_5_disam_list.pkl 
/home/ann/fiz-ddb/notebook/data/5fold-mistralv3/flattened_test_new_candi_5_disam_list.pkl.log

2025-03-10 14:52:02.930949

Output()

[52.30] (flattened_test_new_candi_5_disam_list.pkl)

/home/ann/fiz-ddb/notebook/data/vars/flattened_test_new_term_1_disam_list.pkl 
/home/ann/fiz-ddb/notebook/data/5fold-mistralv3/flattened_test_new_term_1_disam_list.pkl.log

2025-03-10 14:56:10.010384

Output()

[61.89] (flattened_test_new_term_1_disam_list.pkl)

/home/ann/fiz-ddb/notebook/data/vars/flattened_test_new_term_2_disam_list.pkl 
/home/ann/fiz-ddb/notebook/data/5fold-mistralv3/flattened_test_new_term_2_disam_list.pkl.log

2025-03-10 15:00:10.826577

Output()

[63.54] (flattened_test_new_term_2_disam_list.pkl)

/home/ann/fiz-ddb/notebook/data/vars/flattened_test_new_term_3_disam_list.pkl 
/home/ann/fiz-ddb/notebook/data/5fold-mistralv3/flattened_test_new_term_3_disam_list.pkl.log

2025-03-10 15:03:39.060293

Output()

[66.67] (flattened_test_new_term_3_disam_list.pkl)

/home/ann/fiz-ddb/notebook/data/vars/flattened_test_new_term_4_disam_list.pkl 
/home/ann/fiz-ddb/notebook/data/5fold-mistralv3/flattened_test_new_term_4_disam_list.pkl.log

2025-03-10 15:06:58.700220

Output()

[52.58] (flattened_test_new_term_4_disam_list.pkl)

/home/ann/fiz-ddb/notebook/data/vars/flattened_test_new_term_5_disam_list.pkl 
/home/ann/fiz-ddb/notebook/data/5fold-mistralv3/flattened_test_new_term_5_disam_list.pkl.log

2025-03-10 15:11:23.918531

Output()

[55.56] (flattened_test_new_term_5_disam_list.pkl)

>[50.60] (flattened_test_new_candi_1_disam_list.pkl)
>
>[52.15] (flattened_test_new_candi_2_disam_list.pkl)
>
>[50.54] (flattened_test_new_candi_3_disam_list.pkl)
>
>[54.88] (flattened_test_new_candi_4_disam_list.pkl)
>
>[52.30] (flattened_test_new_candi_5_disam_list.pkl)
>
>[61.89] (flattened_test_new_term_1_disam_list.pkl)
>
>[63.54] (flattened_test_new_term_2_disam_list.pkl)
>
>[66.67] (flattened_test_new_term_3_disam_list.pkl)
>
>[52.58] (flattened_test_new_term_4_disam_list.pkl)
>
>[55.56] (flattened_test_new_term_5_disam_list.pkl)